src/uct/cuda/cuda_ipc/cuda_ipc_ep.c

```diff
@@ -18,29 +29,217 @@ static UCS_CLASS_INIT_FUNC(uct_cuda_ipc_ep_t, uct_iface_t *tl_iface,
     uct_cuda_ipc_iface_t *iface = ucs_derived_of(tl_iface, uct_cuda_ipc_iface_t);
 
     UCS_CLASS_CALL_SUPER_INIT(uct_base_ep_t, &iface->super);
+    sglib_hashed_uct_cuda_ipc_rem_seg_t_init(self->rem_segments_hash);
 
     return UCS_OK;
 }
 
 static UCS_CLASS_CLEANUP_FUNC(uct_cuda_ipc_ep_t)
 {
+    struct sglib_hashed_uct_cuda_ipc_rem_seg_t_iterator iter;
+    uct_cuda_ipc_rem_seg_t *remote_seg;
+    ucs_status_t status;
+
+    for (remote_seg = sglib_hashed_uct_cuda_ipc_rem_seg_t_it_init(&iter, self->rem_segments_hash);
+         remote_seg != NULL;
+         remote_seg = sglib_hashed_uct_cuda_ipc_rem_seg_t_it_next(&iter)) {
+            sglib_hashed_uct_cuda_ipc_rem_seg_t_delete(self->rem_segments_hash,
+                                                       remote_seg);
+            status = UCT_CUDADRV_FUNC(cuIpcCloseMemHandle(remote_seg->d_bptr));
+            if (UCS_OK != status) {
+                ucs_fatal("cuIpcCloseMemHandle failed. d_ptr :%p",
+                          (void *) remote_seg->d_bptr);
+            }
+            ucs_free(remote_seg);
+    }
 }
 
 UCS_CLASS_DEFINE(uct_cuda_ipc_ep_t, uct_base_ep_t)
 UCS_CLASS_DEFINE_NEW_FUNC(uct_cuda_ipc_ep_t, uct_ep_t, uct_iface_t*,
                           const uct_device_addr_t *, const uct_iface_addr_t *);
 UCS_CLASS_DEFINE_DELETE_FUNC(uct_cuda_ipc_ep_t, uct_ep_t);
 
-ucs_status_t uct_cuda_ipc_ep_get_zcopy(uct_ep_h tl_ep, const uct_iov_t *iov, size_t iovcnt,
-                                       uint64_t remote_addr, uct_rkey_t rkey,
-                                       uct_completion_t *comp)
+#define uct_cuda_ipc_trace_data(_remote_addr, _rkey, _fmt, ...)     \
+    ucs_trace_data(_fmt " to %"PRIx64"(%+ld)", ## __VA_ARGS__, (_remote_addr), \
+                   (_rkey))
+
+#define UCT_CUDA_IPC_ZERO_LENGTH_POST(len)                      \
+    do {                                                        \
+        if (0 == len) {                                         \
+            ucs_trace_data("Zero length request: skip it");     \
+            return UCS_OK;                                      \
+        }                                                       \
+    } while(0);
+
+void *uct_cuda_ipc_ep_attach_rem_seg(uct_cuda_ipc_ep_t *ep,
+                                     uct_cuda_ipc_iface_t *iface,
+                                     uct_cuda_ipc_key_t *rkey)
 {
-    return UCS_OK;
+    uct_cuda_ipc_rem_seg_t *rem_seg, search;
+    ucs_status_t status;
+
+    /* Are all other members of *search* zeroed out? or ignored?*/
+    search.ph = rkey->ph;
+    rem_seg = sglib_hashed_uct_cuda_ipc_rem_seg_t_find_member(ep->rem_segments_hash, &search);
```

Comment:
May be we can take out whole hashing logic out of this PR. will push later separately. what do you think? for this PR we can open and close handle each time. wht do you think?

Reply:
We may have to do that. We're still about 580 lines now.

Reply:
I've removed the hashing logic now.